File: backend/app/schemas/notifications.py
from typing import Literal

from pydantic import BaseModel, Field

ChannelKind = Literal["discord", "ntfy", "gotify"]


class ChannelRead(BaseModel):
    id: int
    kind: ChannelKind
    name: str
    enabled: bool
    events: list[str]
    # Adresse renvoyée uniquement pour Gotify : celle de Discord (webhook) et
    # de ntfy (nom du sujet) sont des secrets.
    url: str | None = None
    url_set: bool
    token_set: bool


class ChannelWrite(BaseModel):
    kind: ChannelKind
    name: str = Field(max_length=40)
    enabled: bool = True
    events: list[str] = []
    # À la création : obligatoire. À la mise à jour : vide = valeur conservée.
    url: str | None = Field(default=None, max_length=2048)
    token: str | None = Field(default=None, max_length=256)


class ChannelTestResult(BaseModel):
    ok: bool
    # Message d'erreur réduit (code HTTP ou type d'erreur), jamais l'URL.
    error: str | None = None

File: frontend/src/types/notifications.ts
export type ChannelKind = "discord" | "ntfy" | "gotify"

// Même liste que le backend (services/notifications.NOTIFICATION_EVENTS).
export const NOTIFICATION_EVENTS = [
  "scan_completed",
  "scan_failed",
  "orphan_detected",
  "duplicate_detected",
  "non_hardlink_detected",
  "import_failed_detected",
  "stalled_download_detected",
  "delete_selection",
  "import_retry",
  "cascade_delete",
  "hardlink_repair",
  "cross_seed_search",
  "automation",
] as const

export type NotificationEvent = (typeof NOTIFICATION_EVENTS)[number]

export interface NotificationChannel {
  id: number
  kind: ChannelKind
  name: string
  enabled: boolean
  events: NotificationEvent[]
  // Renseignée uniquement pour Gotify : les adresses Discord et ntfy sont des secrets.
  url: string | null
  url_set: boolean
  token_set: boolean
}

export interface NotificationChannelWrite {
  kind: ChannelKind
  name: string
  enabled: boolean
  events: NotificationEvent[]
  // Vide à la modification : conserve l'adresse enregistrée.
  url?: string
  token?: string
}

export interface ChannelTestResult {
  ok: boolean
  error: string | null
}
